File: packages/shared/src/betting.ts
import type { BotDifficulty } from "./game";

/** USD bet amount per difficulty tier (applies to both ranked and casual non-beginner) */
export const BET_AMOUNTS_USD: Record<BotDifficulty, number> = {
  beginner: 10,
  intermediate: 25,
  advanced: 100,
  master: 500,
};

/** House cut in basis points (1000 = 10%) */
export const HOUSE_CUT_BPS = 1000;

/** Casual beginner is free to play */
export const CASUAL_BEGINNER_FREE = true;

/** Donation limits for casual beginner (USD) */
export const DONATION_MIN_USD = 1;
export const DONATION_MAX_USD = 5;

/** Deposit timeout in milliseconds (2 minutes) */
export const DEPOSIT_TIMEOUT_MS = 120_000;

export type GameMode = "ranked" | "casual";

/** Player-selected stake level for a match. */
export type StakeLevel = "free" | "bet";

export interface BetInfo {
  amountWei: string;
  amountUsd: number;
  difficulty: BotDifficulty;
  mode: GameMode;
  contractAddress: string;
  gameId: string;
}

export interface BetSettlement {
  winnerAddress: string | null; // null for draw
  payoutWei: string;
  houseCutWei: string;
  txHash: string;
}

/** Returns true if this mode+difficulty combo is free (no bet required) */
export function isFreeGame(mode: GameMode, difficulty: BotDifficulty): boolean {
  return mode === "casual" && difficulty === "beginner";
}

/** Returns the USD bet amount for a difficulty tier. */
export function getBetAmountUsd(difficulty: BotDifficulty): number {
  return BET_AMOUNTS_USD[difficulty];
}

/** Returns true if the player chose a free stake. */
export function isFreeStake(stake: StakeLevel): boolean {
  return stake === "free";
}

/** Calculate payout amounts from total pot */
export function calculatePayout(totalPotWei: bigint): {
  winnerPayout: bigint;
  houseCut: bigint;
} {
  const houseCut = (totalPotWei * BigInt(HOUSE_CUT_BPS)) / 10000n;
  const winnerPayout = totalPotWei - houseCut;
  return { winnerPayout, houseCut };
}

File: checkker_mobile/lib/models/betting.dart
import 'game.dart';

const Map<BotDifficulty, int> betAmountsUsd = {
  BotDifficulty.beginner: 10,
  BotDifficulty.intermediate: 25,
  BotDifficulty.advanced: 100,
  BotDifficulty.master: 500,
};

const int houseCutBps = 1000;
const bool casualBeginnerFree = true;
const int donationMinUsd = 1;
const int donationMaxUsd = 5;
const int depositTimeoutMs = 120000;

enum GameMode { ranked, casual }

GameMode parseGameMode(String m) => m == 'ranked' ? GameMode.ranked : GameMode.casual;

class BetInfo {
  final String amountWei;
  final double amountUsd;
  final BotDifficulty difficulty;
  final GameMode mode;
  final String contractAddress;
  final String gameId;

  const BetInfo({
    required this.amountWei, required this.amountUsd,
    required this.difficulty, required this.mode,
    required this.contractAddress, required this.gameId,
  });

  factory BetInfo.fromJson(Map<String, dynamic> json) {
    return BetInfo(
      amountWei: json['amountWei'] as String? ?? '0',
      amountUsd: (json['amountUsd'] as num?)?.toDouble() ?? 0.0,
      difficulty: parseBotDifficulty(json['difficulty'] as String? ?? 'beginner'),
      mode: parseGameMode(json['mode'] as String? ?? 'casual'),
      contractAddress: json['contractAddress'] as String? ?? '',
      gameId: json['gameId'] as String? ?? '',
    );
  }
}

class BetSettlement {
  final String? winnerAddress;
  final String payoutWei;
  final String houseCutWei;
  final String txHash;

  const BetSettlement({
    this.winnerAddress, required this.payoutWei,
    required this.houseCutWei, required this.txHash,
  });
}

bool isFreeGame(GameMode mode, BotDifficulty difficulty) {
  return mode == GameMode.casual && difficulty == BotDifficulty.beginner;
}
